plugin/evm/limit_order.go

```diff
@@ -72,13 +73,20 @@ func (lop *limitOrderProcesser) RunMatchingEngine() {
 	if len(longOrders) == 0 || len(shortOrders) == 0 {
 		return
 	}
-	for _, longOrder := range longOrders {
-		for j, shortOrder := range shortOrders {
-			if longOrder.Price == shortOrder.Price && longOrder.BaseAssetQuantity == (-shortOrder.BaseAssetQuantity) {
-				err := lop.limitOrderTxProcessor.ExecuteMatchedOrdersTx(*longOrder, *shortOrder)
+	for i := 0; i < len(longOrders); i++ {
+		for j := 0; j < len(shortOrders); j++ {
+			if getUnFilledBaseAssetQuantity(longOrders[i]) == 0 {
+				break
+			}
+			if getUnFilledBaseAssetQuantity(shortOrders[j]) == 0 {
+				continue
+			}
+			if longOrders[i].Price == shortOrders[j].Price {
+				fillAmount := math.Abs(math.Min(float64(getUnFilledBaseAssetQuantity(longOrders[i])), float64(-(getUnFilledBaseAssetQuantity(shortOrders[j])))))
+				err := lop.limitOrderTxProcessor.ExecuteMatchedOrdersTx(longOrders[i], shortOrders[j], uint(fillAmount))
 				if err == nil {
-					shortOrders = append(shortOrders[:j], shortOrders[j+1:]...)
-					break
+					longOrders[i].FilledBaseAssetQuantity = longOrders[i].FilledBaseAssetQuantity + int(fillAmount)
+					shortOrders[j].FilledBaseAssetQuantity = shortOrders[j].FilledBaseAssetQuantity - int(fillAmount)
```

Comment:
If we update them during block production, won't we need to reset them just like we `purgeLocalTx` ?

Reply:
I have changed getLongOrders() and getShortOrders() to return limit_orders by value instead of instead of by reference. Earlier it was returning pointer to LimitOrder in memory_database.
So any changes here is var longOrders and shortOrders are in local copy in that method.
The final update in filledBaseAssetQuantity happens after tx parsing in limit_order_tx_processor:97